include/net/nrf_cloud_coap.h

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef NRF_CLOUD_COAP_H_
+#define NRF_CLOUD_COAP_H_
+
+/** @file nrf_cloud_coap.h
+ * @brief Module to provide nRF Cloud CoAP API
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_agps.h>
+#include <net/nrf_cloud_pgps.h>
+#include <zephyr/net/coap_client.h>
+
+/**
+ * @defgroup nrf_cloud_coap nRF CoAP API
+ * @{
+ */
+
+/* Transport functions */
+/**@brief Initialize nRF Cloud CoAP library.
+ *
+ * @return 0 if initialization was successful, otherwise, a negative error number.
+ */
+int nrf_cloud_coap_init(void);
```

Comment:
A lot of LTE networks have quite short NATs / timeouts. So in the common case I expect that the socket needs to be opened/close for every send to avoid timing out. Keeping the socket opened for more than some 60 seconds will likely cause a timeout. Patching in @stig-bjorlykke as I belive this is what they do in carrier lib. We enable queue mode in ATv2 lwm2m builds which essenatially does this.

Having this in mind it makes sense to have (`_connect` / `_ disconnect`) flavors that can be called for every update to some resource.

Atleast that is very intuitive from an application standpoint. And Is there any reason why a user might want to be connected, but not authenticated? Doesn't the act of calling connect imply that the application wants to do some resource manipulation and thus needs to be authenticated?

IMO this would be a clean abstraction:
- `nrf_cloud_coap_init()` - called once with other initialization functions in the app.
- `nrf_cloud_coap_connect()`
- `nrf_cloud_coap_close()`

If nessecary to keep authenticaion and connect separate, the authentication API could still be publically exposed. But, there could be a Kconfig option that makes the connect API handle authentication automatically.

Reply:
@simensrostad yes, obviously, NAT is an issue.  That's why CID was added to DTLS, and why using mfw_1.3.5 is important.

I'll consider your input and Tommi's and adjust the abstraction, but it is not necessary to open/close every send.

Reply:
Thanks, what is the recommended approach in case the update interval exceeds NAT?

Reply:
It's almost "don't worry about it."  My sample has an option "DELAY_INTERPACKET_PERIOD" which increases the time between CoAP interactions every time, starting with APP_COAP_SEND_INTERVAL_MS seconds (default 20 seconds) and remaining eventually at APP_COAP_SEND_INTERVAL_MS * APP_COAP_INTERVAL_LIMIT.  I've had it run a long time, up to 20 minutes between transmissions, and there is no trouble continuing to communicate with CID enabled.  Sometimes the LTE connection drops, so you need to handle that error and reconnect.

It has been a while since I tested this specifically, so I will run again today and update this comment accordingly.

Reply:
It still works.  It remains in RRC idle mode and CSCON 0 until the next time to do a transfer.

Reply:
I've renamed the various functions.  Note that the original design of this came from the nrf9160/coap_client, which creates the socket and connects to it in the client_init() function.  But it's fine to move this to nrf_cloud_coap_connect().